File: frontend/lib/phone.ts
/**
 * Client-side Morocco mobile normalization and validation.
 * Mirrors backend app/services/phone.py
 */

const MA_PATTERNS = [
  /^\+212[67](\d{8})$/,
  /^00212[67](\d{8})$/,
  /^212[67](\d{8})$/,
  /^0[67](\d{8})$/,
  /^[67](\d{8})$/,
];

export function normalizeMaPhone(raw: string): {
  e164: string;
  digits: string;
} {
  const cleaned = raw.replace(/[\s\-\(\)]/g, "").trim();

  for (const pattern of MA_PATTERNS) {
    const m = cleaned.match(pattern);
    if (m) {
      const local8 = m[1];
      const prefix = local8[0];
      const digits = `212${prefix}${local8}`;
      return { e164: `+${digits}`, digits };
    }
  }

  throw new Error("رقم الجوال يجب أن يكون رقماً مغربياً صحيحاً يبدأ بـ 06 أو 07");
}

export function isValidMaPhone(raw: string): boolean {
  try {
    normalizeMaPhone(raw);
    return true;
  } catch {
    return false;
  }
}

/** @deprecated */
export const normalizeKsaPhone = normalizeMaPhone;
/** @deprecated */
export const isValidKsaPhone = isValidMaPhone;

File: backend/app/services/phone.py
import re


def normalize_ma_phone(raw: str) -> tuple[str, str]:
    """
    Normalize a Morocco mobile phone number.
    Returns (e164, digits) or raises ValueError.

    e164:   +2126xxxxxxxx or +2127xxxxxxxx
    digits: 2126xxxxxxxx (used for CAPI hashing)
    """
    cleaned = re.sub(r"[\s\-\(\)]", "", raw.strip())

    patterns = [
        r"^\+212([67])(\d{8})$",
        r"^00212([67])(\d{8})$",
        r"^212([67])(\d{8})$",
        r"^0([67])(\d{8})$",
        r"^([67])(\d{8})$",
    ]

    prefix = None
    local_8 = None
    for pattern in patterns:
        m = re.match(pattern, cleaned)
        if m:
            prefix = m.group(1)
            local_8 = m.group(2)
            break

    if local_8 is None or prefix is None:
        raise ValueError(
            "رقم الجوال يجب أن يكون رقماً مغربياً صحيحاً يبدأ بـ 06 أو 07"
        )

    digits = f"212{prefix}{local_8}"
    e164 = f"+{digits}"
    return e164, digits


# Backward-compatible alias for imports
normalize_ksa_phone = normalize_ma_phone
